src/unstated.js

```diff
@@ -13,8 +13,20 @@ export class Container<State: {}> {
     this._listeners = [];
   }
 
-  setState(state: $Shape<State>) {
-    this.state = Object.assign({}, this.state, state);
+  setState(state: $Shape<State> | (State => $Shape<State>)): void {
+    const prevState = this.state;
+    const newState = typeof state === 'function' ? state(prevState) : state;
+
+    /**
+     * allow either early return or ternary return of previous state to
+     * propagate correctly through component tree by not calling _listeners
+     * and leaving old state in place.
+     */
+    if (newState == null || newState === prevState) {
```

Comment:
I think returning early if a shallow comparison fails would be a breaking change, and incompatible with React's `setState` behavior. React docs discourage returning mutated state from the `setState` callback (or passing a mutated state object to `setState`), but neither is prohibited. From [the React docs](https://reactjs.org/docs/react-component.html#setstate):

> `setState()` will always lead to a re-render unless `shouldComponentUpdate()` returns false.

My second concern is that React emits its own development-time console warnings for passing `null` to `setState`, and by returning silently we might prevent a developer from realizing they have made a mistake.

For both of those reasons, I think it might be best to eliminate the conditional early return altogether.

Reply:
So in other words, I would think about rolling back ff77c96e775d39d7746773c67e37f1d2e97afba6.

Reply:
> setState() will always lead to a re-render unless shouldComponentUpdate() returns false.

I think this is not true with react@16. I don't use this feature but returning `null` is supposed to prevent scheduling a re-render.

What we might need to introduce here is 2 distinct paths for objects & function styles. Because of the `null` issue we cannot combine both and use `newState` in following lines, it should be restructured to something like:
```js
  setState(state: $Shape<State> | (State => $Shape<State>)): void {
    const prevState = this.state;
    let newState
    if (typeof state === 'function') {
      newState = state(prevState)
      
      if (newState === null) {
        return;
      }
    } else {
      newState = state

      if (process.env.NODE_ENV !== 'production' && newState === null) {
        console.warn('write a nice dev warning here')
      }
    }

    if (newState === prevState) {
      return
    }

    this.state = Object.assign({}, prevState, newState);
    this._listeners.forEach(fn => fn());
  }
```

Does it make sense to u?

Reply:
Yeah I understand how to get there code-wise, I'd just like to have @jamiebuilds's input on what the desired behavior is.

Reply:
@Andarist Ok I've investigated the actual behavior and this is what I found.
(Codepen using React 15: https://codepen.io/anon/pen/QQYRBx)
(JSFiddle with the same code using React 16: https://jsfiddle.net/0ehzq82o/)
1. React 15 **warns** if `null` or `undefined` is passed to `setState` as the first argument. React 16 **does not warn**.
2. Both React 15 and React 16 **do not warn** if `null` or `undefined` is returned from a `setState` updater function (I didn't realize this!).
3. In all cases, a `null` or `undefined` state update of any type will not alter the existing state.

Since this library is intended to support both React 15 and React 16, my instinct says it's best to defer to React's implementation of `setState` rather than trying to reimplement the error handling.

Pros:

Reply:
Also @Andarist the code example you shared still says not to update if `newState === prevState` - which is not a valid reason React would not update.

Reply:
> Since this library is intended to support both React 15 and React 16, my instinct says it's best to defer to React's implementation of setState rather than trying to reimplement the error handling.

How does error handling matter here?

> Also @Andarist the code example you shared still says not to update if newState === prevState - which is not a valid reason React would not update.

Right, I got used too much to thinking in terms of `PureComponent`s 😄

Reply:
> How does error handling matter here?

I think I misread the part where you `console.warn('write a nice dev warning here')` - I thought you were also returning. You *are* returning if the update is `null` for the updater function, but since you correctly pointed out React doesn't (and perhaps has never) emitted warnings for this, I suppose it's probably fine. 🙂

Might be good to check for both `undefined` *and* `null` though, since I think they're treated the same by `setState`.

Reply:
I've added warning based on your comment: 
> My second concern is that React emits its own development-time console warnings for passing null to setState, and by returning silently we might prevent a developer from realizing they have made a mistake.

> Might be good to check for both undefined and null though, since I think they're treated the same by setState.

Yeah, needs to be re-checked against React's implementation though.

Reply:
Right, I understand. My suggestion was to simply pass through to React and let it report errors on its own. Either way, I'm happy as long as this isn't preventing updates that would work in React! 😄

Reply:
> My suggestion was to simply pass through to React and let it report errors on its own.

That's actually not possible, unstated's `setState` has nothing to do with React's `setState`. React cannot link those 2 together, especially that we actually setState with some DUMMY_STATE object in React, just to cause a re-render.

Reply:
Yeah since this is already only conceptually similar to setState, I'm not convinced we should be following React to the letter vs doing the principally right thing. Their API tradeoffs aren't necessarily the same as those of a pure state manager. I'll update tomorrow, and hopefully get some time in to write those tests.

Reply:
I want it to match the React behaviour as closely as possible. Refactoring component state into container state should be near seamless

Reply:
Wow my mistake! I was operating under the impression the `Subscribe` component's `onUpdate()` subscription callback was receiving the updated state and passing it to `setState()`. I should have read more closely. 🙂 

@StevenLangbroek Given the above discussion I concede it makes sense to have custom error reporting and perhaps disregard what React does or does not report as a mistake.

However, the documentation for Unstated is currently written without any specification for `setState()`. The implication is that `setState` should behave the same way you would expect already, having worked with React. That means:

If there's an intent to

Reply:
@jamiebuilds I missed your reply, but I guess I should have just let your comment sit. 😄 I appreciate that sentiment - it's my favorite thing about this library.

Reply:
> ```js
>   if (process.env.NODE_ENV !== 'production' && newState === null) {
>         console.warn('write a nice dev warning here')
>       }
> ```

Hi @Andarist, can we make one constant file and manage this kind of string `production` ? because in code we are using this kind of thing repeatedly and so instead of doing repetitively we can handle it in one file use.